Clearing reading list item list on archive status toggle.

thomasstep opened the pull request:
<!--- Prepend PR title with [WIP] if work in progress. Remove when ready for review. -->

<!--- For a timely review/response, please avoid force-pushing additional commits if your PR already received reviews or comments -->

## What type of PR is this? (check all applicable)

- [ ] Refactor
- [ ] Feature
- [✅] Bug Fix
- [ ] Documentation Update

## Description
I believe the cached search results from Algolia were causing the reading list status views to not properly update.

## Related Tickets & Documents
[Algolia's documentation about clearing cache](https://www.algolia.com/doc/api-client/advanced/cache-browser-only/javascript/?language=javascript)
#3217 

## Mobile & Desktop Screenshots/Recordings (if there are UI changes)
N/A

## Added to documentation?

- [ ] docs.dev.to
- [ ] readme
- [✅] no documentation needed

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/thepracticaldev/dev.to?pullRequest=4952) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/thepracticaldev/dev.to?pullRequest=4952) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/thepracticaldev/dev.to?pullRequest=4952) it.</sub>

thomasstep commented:
Hello @rhymes 

> Just a question: could this be called in the `componentDidMount` or somewhere not just related to clicking the archive button?

It could also be called in `searchableItemList.js` before a search is performed. I haven't taken the time to see what else would be affected by that though. It could potentially lead to slower response times thought since it would actually be reaching out to Algolia each time the cache is cleared.

> Do you think the other one could be solved by this as well?

Do you know all the components the Reading List count is being calculated in? I tried to looking for it outside of readingList and could not find it, but I did not look very hard. I have not tried to recreate the issue in #1061 but it might be fixed by this. I can look into it more if you want.

rhymes commented:
> It could also be called in searchableItemList.js before a search is performed. I haven't taken the time to see what else would be affected by that though. It could potentially lead to slower response times thought since it would actually be reaching out to Algolia each time the cache is cleared.

I would start calling it from the `readingList` component, `searchableItemList` is also used in the `history` component, and there is not needed, since that list is only additive.

BTW upon [further reading](https://www.algolia.com/doc/api-client/advanced/cache-browser-only/javascript/?language=javascript) I don't think it would solve the other issue we have:

> It’s stored as a simple JavaScript object in memory, so the cache is automatically reset whenever the page is refreshed.

because the other bug persists on page refresh.

> Do you know all the components the Reading List count is being calculated in? I tried to looking for it outside of readingList and could not find it, but I did not look very hard.

No, only there. 

> I have not tried to recreate the issue in #1061 but it might be fixed by this. I can look into it more if you want.

No sorry for bringing that up, upon further inspection I think it's a different issue, it has probably more to do with server side caching and the deletion of reactions than it has to do with the calling of Algolia in JS.

thomasstep commented:
@rhymes thank you for the comments. Is there anything that I should adjust in this PR? I'm not sure what the normal life cycle is for a PR, so if there is anything I need to do that I'm not please let me know.

thomasstep commented:
@rhymes That was my fault for not completely understanding the issue until now. I was trying to solve the issue of an archived article still showing up in the reading list and vice versa but not the second part which is a deleted article still showing up. How did you get your environment into a state with an article that is not in the DB but still in Algolia?

I created an article through the app, added it to my reading list, deleted it (in a separate tab), and then tried to archive the deleted article. I see the 404 error that you're talking about; however, for me the article is removed from the reading list immediately (I believe due to line 107 in `readingList.jsx`) and on refresh. 

https://github.com/thepracticaldev/dev.to/blob/585d672742a8f5b2e7dd563b67cec9398d25d44f/app/javascript/readingList/readingList.jsx#L107

I don't understand the system/app as well as you do, but what would have to happen in order for this situation to occur and what would the use case be?

Thank you for all the replies and effort that you're putting into this. I appreciate the help!

mstruve commented:
Hi @thomasstep! 

Thank you so much for the PR and for taking the leap and contributing to DEV! We are super happy to have your help 😄
 
With that said, we are in the process of going through and cleaning up some of our PR history to help us better prioritize what we need to be focusing on. This change is no longer needed since we have moved off Algolia and on to Elasticsearch so I am going to close this PR. 

Thanks again for contributing! Ping us anytime if you have questions!
